File: py/bands.py
"""Band generator — concentric outward fill from a seed mask.
See spec/bands.md.  Mirrors rs/src/bands.rs exactly for delta-0 diffs.
"""

from __future__ import annotations
import numpy as np
from dataclasses import dataclass


@dataclass
class Band:
    mask: np.ndarray   # (H, W) bool
    white: bool        # True = white, False = black


def growth_to_thickness(growth_pct: float) -> int:
    """Map growth% [0,100] -> thickness_px [1,40] with round-half-up."""
    x = 1.0 + (growth_pct / 100.0) * 39.0
    return int(x + 0.5)  # floor(x + 0.5) = round-half-up (matches Rust)


def dilate_chebyshev(mask: np.ndarray, r: int) -> np.ndarray:
    """Chebyshev (box) dilation by r, two-pass 1D sliding window.
    Matches rs/src/bands.rs::dilate_chebyshev exactly.
    mask: (H, W) bool ndarray.
    """
    if r == 0:
        return mask.copy()
    h, w = mask.shape

    # ---- Horizontal pass ----
    tmp = np.zeros((h, w), dtype=bool)
    for x in range(w):
        x_lo = max(0, x - r)
        x_hi = min(w - 1, x + r)
        tmp[:, x] = np.any(mask[:, x_lo:x_hi + 1], axis=1)

    # ---- Vertical pass ----
    result = np.zeros((h, w), dtype=bool)
    for y in range(h):
        y_lo = max(0, y - r)
        y_hi = min(h - 1, y + r)
        result[y, :] = np.any(tmp[y_lo:y_hi + 1, :], axis=0)

    return result


def generate_bands(
    seed: np.ndarray,      # (H, W) bool — opaque pixels of seed layer
    thickness_px: int,
    fill_pct: float,       # 0..100
) -> list[Band]:
    """Generate bands outward from seed, innermost first (index 0 = white)."""
    h, w = seed.shape
    total = w * h
    coverage = seed.copy()
    coverage_count = int(coverage.sum())
    fill_threshold = int(np.ceil(fill_pct / 100.0 * total))
    bands: list[Band] = []

    while True:
        dilated  = dilate_chebyshev(coverage, thickness_px)
        new_ring = dilated & ~coverage

        new_count = int(new_ring.sum())
        if new_count == 0:
            break  # stop condition 2: canvas saturated

        is_white = len(bands) % 2 == 0  # white first
        bands.append(Band(mask=new_ring.copy(), white=is_white))

        coverage      |= new_ring
        coverage_count += new_count

        if coverage_count >= fill_threshold:
            break  # stop condition 1: fill% reached

    return bands

File: rs/src/bands.rs
//! Band generator — concentric outward fill from a seed mask.
//! See spec/bands.md.

/// One generated band: which pixels it covers and what colour.
pub struct Band {
    pub mask: Vec<bool>, // W*H, row-major; true = this band's pixel
    pub white: bool,     // true = white (#fff), false = black (#000)
}

/// Map growth percentage [0,100] to thickness in pixels [1,40].
/// Uses round-half-up so both Python and Rust agree on every integer growth%.
pub fn growth_to_thickness(growth_pct: f32) -> usize {
    let x = 1.0_f32 + (growth_pct / 100.0) * 39.0;
    (x + 0.5).floor() as usize
}

/// Generate bands outward from `seed` (boolean W×H mask, row-major).
/// Returns bands ordered **innermost first** (index 0 = band closest to seed,
/// white). Thickness `thickness_px` is the Chebyshev dilation radius per
/// iteration. Stops when coverage ≥ fill_pct% or the canvas saturates.
pub fn generate_bands(
    seed: &[bool],
    width: usize,
    height: usize,
    thickness_px: usize,
    fill_pct: f32,
) -> Vec<Band> {
    let total = width * height;
    let mut coverage: Vec<bool> = seed.to_vec();
    let mut coverage_count: usize = coverage.iter().filter(|&&b| b).count();
    let mut bands: Vec<Band> = Vec::new();
    let fill_threshold = (fill_pct / 100.0 * total as f32).ceil() as usize;

    loop {
        let dilated  = dilate_chebyshev(&coverage, width, height, thickness_px);
        let new_ring: Vec<bool> = dilated.iter().zip(coverage.iter())
            .map(|(&d, &c)| d && !c)
            .collect();

        let new_count = new_ring.iter().filter(|&&b| b).count();
        if new_count == 0 {
            break; // stop condition 2: canvas saturated
        }

        let is_white = bands.len().is_multiple_of(2); // white first
        bands.push(Band { mask: new_ring.clone(), white: is_white });

        for (c, n) in coverage.iter_mut().zip(new_ring.iter()) {
            if *n { *c = true; }
        }
        coverage_count += new_count;

        if coverage_count >= fill_threshold {
            break; // stop condition 1: fill% reached
        }
    }

    bands
}

// ---------------------------------------------------------------------------
// Chebyshev (box) dilation — two-pass 1D sliding window, O(W·H) per call.
// Matches py/bands.py's dilate_chebyshev exactly.
// ---------------------------------------------------------------------------

pub fn dilate_chebyshev(mask: &[bool], w: usize, h: usize, r: usize) -> Vec<bool> {
    if r == 0 {
        return mask.to_vec();
    }

    // ---- Horizontal pass ----
    let mut tmp = vec![false; w * h];
    for y in 0..h {
        let base = y * w;
        // Initialize count for x=0: window covers [0 .. min(r, w-1)]
        let mut count: usize = (0..=r.min(w.saturating_sub(1)))
            .filter(|&nx| mask[base + nx])
            .count();
        tmp[base] = count > 0;
        for x in 1..w {
            // Right edge enters window
            if x + r < w && mask[base + x + r] { count += 1; }
            // Left edge leaves window
            if x > r && mask[base + x - r - 1] { count -= 1; }
            tmp[base + x] = count > 0;
        }
    }

    // ---- Vertical pass ----
    let mut result = vec![false; w * h];
    for x in 0..w {
        // Initialize count for y=0: window covers [0 .. min(r, h-1)]
        let mut count: usize = (0..=r.min(h.saturating_sub(1)))
            .filter(|&ny| tmp[ny * w + x])
            .count();
        result[x] = count > 0;
        for y in 1..h {
            if y + r < h && tmp[(y + r) * w + x] { count += 1; }
            if y > r     && tmp[(y - r - 1) * w + x] { count -= 1; }
            result[y * w + x] = count > 0;
        }
    }

    result
}
